db/src/types.rs

```diff
@@ -0,0 +1,251 @@
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![allow(dead_code)]
+
+use std::collections::{BTreeMap};
+
+/// Core types defining a Mentat knowledge base.
+///
+/// At its core, Mentat maintains a set of assertions of the form entity-attribute-value (EAV).  The
+/// assertions conform to a schema where-by the given attribute constrains the associated value/set
+/// of associated values.
+///
+/// ## Assertions
+///
+/// Mentat assertions are represented as rows in the `datoms` SQLite table, and each Mentat row
+/// representing an assertion are tagged with a numeric representation of :db/valueType.
+///
+/// The tag is used to limit queries, and therefore is placed carefully in the relevant indices to
+/// allow searching numeric longs and doubles quickly.  The tag is also used to convert SQLite
+/// values to the correct Mentat value type on query egress.
+///
+/// ## Entities and entids
+///
+/// A Mentat entity is represented by a *positive* integer.  (This agrees with Datomic.)  We call
+/// such a positive integer an *entid*.
+///
+/// ## Partitions
+///
+/// Datomic partitions the entid space in order to separate core knowledge base entities required
+/// for the healthy function of the system from user-defined entities.  Datomic also partitions in
+/// order to ensure that certain index walks of related entities are efficient.  Mentat follows
+/// suit, partitioning into the following partitions:
+/// * `:db.part/db`, for core knowledge base entities;
+/// * `:db.part/user`, for user-defined entities;
+/// * `:db.part/tx`, for transaction entities.
+/// You almost certainly want to add new entities in the `:db.part/user` partition.
+///
+/// The entid sequence in a given partition is monotonically increasing, although not necessarily
+/// contiguous.  That is, it is possible for a specific entid to have never been present in the
+/// system, even though its predecessor and successor is present.
+
+// #[derive(Debug)]
+// pub enum Error {
+//     RusqliteError(rusqlite::Error),
+//     BadSchemaAssertion(String),
+//     BadBootstrapDefinition(String),
+// }
+
+// impl From<rusqlite::Error> for Error {
+//     fn from(e: rusqlite::Error) -> Error {
+//         Error::RusqliteError(e)
+//     }
+// }
+
+/// Represents one entid in the entid space.
+///
+/// Per https://www.sqlite.org/datatype3.html (see also http://stackoverflow.com/a/8499544), SQLite
+/// stores signed integers up to 64 bits in size.  Since u32 is not appropriate for our use case, we
+/// use i64 rather than manually truncating u64 to u63 and casting to i64 throughout the codebase.
+pub type Entid = i64;
+
+/// The attribute of each Mentat assertion has an :db/valueType constraining the value to a
+/// particular domain.  Mentat recognizes the following :db/valueType values.
```

Comment:
I wouldn't use 'domain' here: attributes have a domain (always ref) and a range (which is what valueType defines). Perhaps "value space" instead.